lib/client/local_proxy_middleware.go

```diff
@@ -220,3 +300,199 @@ func (c *AppCertIssuer) IssueCert(ctx context.Context) (tls.Certificate, error)
 	}
 	return appCert, nil
 }
+
+// LocalCertGenerator is a HTTPS listener that can generate TLS certificates based
+// on SNI during HTTPS handshake.
+type LocalCertGenerator struct {
+	certChecker *CertChecker
+	caPath      string
+
+	mu sync.RWMutex
+	// ca is the certificate authority for signing certificates.
+	ca tls.Certificate
+	// certsByHost is a cache of certs for hosts generated with the local CA.
+	certsByHost map[string]*tls.Certificate
+}
+
+// NewLocalCertGenerator creates a new LocalCertGenerator and listens to the
+// configured listen address.
+func NewLocalCertGenerator(certChecker *CertChecker, caPath string) (*LocalCertGenerator, error) {
+	r := &LocalCertGenerator{
+		certChecker: certChecker,
+		caPath:      caPath,
+	}
+
+	if err := r.ensureValidCA(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return r, nil
+}
+
+// GetCertificate generates and returns TLS certificate for incoming
+// connection. Implements tls.Config.GetCertificate.
+func (r *LocalCertGenerator) GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	if err := r.ensureValidCA(); err != nil {
```

Comment:
what happens here if the CA is reissued? Does the app (e.g. aws cli) throw an error because it doesn't recognize the CA? Or does it lazily load the CA from `AWS_CA_BUNDLE`?

Reply:
It reissues the CA to the same path as `AWS_CA_BUNDLE`, so it should work fine.